src/Toml/BiMap.hs

```diff
@@ -57,79 +58,84 @@ import Control.Arrow ((>>>))
 import Data.Word (Word)
 import Numeric.Natural (Natural)
-import Data.Hashable (Hashable)
-import Text.Read (readMaybe)
-import Data.Time (Day, LocalTime, TimeOfDay, ZonedTime)
+import Text.Read (readEither)
 
-import Toml.Type (AnyValue (..), Value (..), DateTime (..) ,
-                  matchArray, matchBool, matchDouble, matchInteger,
-                  matchText, matchDate, toMArray)
+import Toml.Type (AnyValue (..), DateTime (..), Value (..), matchArray, matchBool, matchDate,
+                  matchDouble, matchInteger, matchText, toMArray)
 
 import qualified Control.Category as Cat
-import qualified Data.Text as T
-import qualified Data.Text.Encoding as T
 import qualified Data.ByteString.Lazy as BL
-import qualified Data.Text.Lazy as TL
-import qualified Data.Text.Lazy.Encoding as TL
-import qualified Data.Set as S
 import qualified Data.HashSet as HS
 import qualified Data.IntSet as IS
 import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as S
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Encoding as TL
 
 ----------------------------------------------------------------------------
 -- BiMap concepts and ideas
 ----------------------------------------------------------------------------
 
 {- | Partial bidirectional isomorphism. @BiMap a b@ contains two function:
 
-1. @a -> Maybe b@
-2. @b -> Maybe a@
+1. @a -> Either e b@
+2. @b -> Either e a@
 -}
-data BiMap a b = BiMap
-    { forward  :: a -> Maybe b
-    , backward :: b -> Maybe a
+data BiMap e a b = BiMap
+    { forward  :: a -> Either e b
+    , backward :: b -> Either e a
     }
 
-instance Cat.Category BiMap where
-    id :: BiMap a a
-    id = BiMap Just Just
+instance Cat.Category (BiMap e) where
+    id :: BiMap e a a
+    id = BiMap Right Right
 
-    (.) :: BiMap b c -> BiMap a b -> BiMap a c
+    (.) :: BiMap e b c -> BiMap e a b -> BiMap e a c
     bc . ab = BiMap
         { forward  =  forward ab >=>  forward bc
         , backward = backward bc >=> backward ab
         }
 
 -- | Inverts bidirectional mapping.
-invert :: BiMap a b -> BiMap b a
+invert :: BiMap e a b -> BiMap e b a
 invert (BiMap f g) = BiMap g f
 
 -- | Creates 'BiMap' from isomorphism.
-iso :: (a -> b) -> (b -> a) -> BiMap a b
-iso f g = BiMap (Just . f) (Just . g)
+iso :: (a -> b) -> (b -> a) -> BiMap e a b
+iso f g = BiMap (Right . f) (Right . g)
 
 -- | Creates 'BiMap' from prism-like pair of functions.
-prism :: (field -> object) -> (object -> Maybe field) -> BiMap object field
-prism review preview = BiMap preview (Just . review)
+prism :: (field -> object) -> (object -> Either error field) -> BiMap error object field
+prism review preview = BiMap preview (Right . review)
+
+----------------------------------------------------------------------------
+-- BiMap error type
+----------------------------------------------------------------------------
+
+data BiMapError = BiMapError
```

Comment:
I guess that the most important part of this PR is the type of this error. I can see the following cases:

1. Name of the data type constructor is not what we expected.
2. Name of the `Value` constructor is not what we expected.
3. Arbitrary textual reason for other cases.

Maybe something else. But I guess we can settle to this 3 constructors for now.